### PR TITLE
docs(ADR): standardize Flag configuration parsing and error handling

### PR DESCRIPTION
**Pull request by aepfli**
relates to: #1690

This is the starting point, i think we should talk about what we want to achieve and how. My simple mind did not manage to come to another option i could propose, but i am open to add more options. please let me know what you think of this.

### CODE REVIEW

File: docs/architecture-decisions/flag_configuration_parsing.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+- [BUG] Inconsistency in file provider and JSON parsing for different languages (#1627)
+- [BUG] [File Provider] Support File Formats (#1689)
+- [FEATURE] Provider option to reject flag set with invalid rules (#1487)
+
+To address these issues, it is essential to establish a consistent and testable framework for flag parsing and handling.
+
+## Requirements
+
+* Utilize schema validation as much as possible, instead of custom logic, on a flag level rather than a configuration level.
+* Define what consistutes a `PARSE_ERROR` versus a `FLAG_NOT_FOUND`.
```
Comment (toddbaert) [inline docs/architecture-decisions/flag_configuration_parsing.md]:
```suggestion
* Define what constitutes a `PARSE_ERROR` versus what prevents the flag set from being updated.
```

File: docs/architecture-decisions/flag_configuration_parsing.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+updated: YYYY-MM-DD
+---
+
+# Proposal for Unified Flag Parsing and Handling
+
+The goal of this proposal is to standardize how flags are parsed and handled across implementations. Currently, there is significant variation in how this is done, stemming from the organic growth of the project and contributions from multiple developers. This proposal aims to rethink and unify the parsing and handling of configurations, particularly for in-process scenarios.
+
+## Background
+
+Over time, the project has grown with contributions from various developers, leading to inconsistencies in how flags are parsed and handled. These inconsistencies have resulted in bugs, unexpected behaviors, and challenges in maintaining the codebase. For example, the file provider and JSON parsing behave differently across languages, and there is no clear definition of what constitutes a parsing error versus a flag-not-found error.
```
Comment (toddbaert) [inline docs/architecture-decisions/flag_configuration_parsing.md]:
```suggestion
Initially we did not have a JSON schema, and validation was done manually after the configuration was parsed; in this case parsing was not consistent at all.
Later, we implemented a basic schema for the flag structure, which could be used to consistently validate the everything but the targeting rules.
Finally, we wrote a full schema for the targeting rules as well, which ensure that all `targeting` was valid.
These inconsistencies have resulted in bugs, divergent behaviors, and challenges in maintaining the codebase. For example, the file provider and JSON parsing behave differently across languages, and there is no clear definition of what constitutes a parsing error versus a flag-not-found error, and what exactly happens when an invalid configuration is loaded.
```

File: docs/architecture-decisions/flag_configuration_parsing.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+- [BUG] [File Provider] Support File Formats (#1689)
+- [FEATURE] Provider option to reject flag set with invalid rules (#1487)
+
+To address these issues, it is essential to establish a consistent and testable framework for flag parsing and handling.
+
+## Requirements
+
+* Utilize schema validation as much as possible, instead of custom logic, on a flag level rather than a configuration level.
+* Define what consistutes a `PARSE_ERROR` versus a `FLAG_NOT_FOUND`.
+* Define if an invalid flag entry within a configuration can invalidate the whole configuration.
```
Comment (toddbaert) [inline docs/architecture-decisions/flag_configuration_parsing.md]: I think a JSON document is either invalid or valid, so we don't have unlimited flexibility here unless we want to do some fairly annoying manipulation of the payload, either validating only 1 flag at a time (this might require breaking single flags into a subschema).

It _might_ be possible to identify the specific sub-entity of an invalid configuration with the validation library, and remove it, but I think that might not be worth it or possible everywhere.

File: docs/architecture-decisions/flag_configuration_parsing.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+* Define which file formats are supported (e.g., JSON, YAML, etc.).
```
Comment (toddbaert) [inline docs/architecture-decisions/flag_configuration_parsing.md]: We should make this consistent, I think. It's pretty trivial to support YAML/JSON conversion in most cases, though there might be some edge cases to consider.

File: docs/architecture-decisions/flag_configuration_parsing.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+* Ensure the framework is testable via the existing testbed/test-harness.
```
Comment (toddbaert) [inline docs/architecture-decisions/flag_configuration_parsing.md]: :100:

File: docs/architecture-decisions/flag_configuration_parsing.md (hunk 3 above)
Comment (aepfli) [reply docs/architecture-decisions/flag_configuration_parsing.md]: we can select subsets for validating entities within the json schema, it will be one validation per flag, but i think it is better than invalidating the whole configuration.

File: docs/architecture-decisions/flag_configuration_parsing.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+## Proposal
+
+### Standardize Flag Parsing and Handling Using Schema Validation (Option 1)
+
+This proposal recommends adopting schema validation as the primary method for parsing and handling flags. The key elements of this approach include:
+
+1. **Schema Validation**: Use schema validation to ensure that flags meet predefined criteria. This reduces the reliance on custom logic and improves consistency across implementations.
+2. **Supported File Formats**: Define and document the file formats that are supported (e.g., JSON, YAML). This aligns with the issue raised in [BUG] [File Provider] Support File Formats (#1689).
+3. **Error Definitions**: Clearly define what constitutes a `PARSE_ERROR` versus a `FLAG_NOT_FOUND` error. Each inconsistent flag should evaluate as a `PARSE_ERROR`, only if the flag is not part of the configuration (or due to other definitions), it should evaluate to a `FLAG_NOT_FOUND`
+4. **Invalid Flag Handling**: A invalid flag configuration should not affect the whole configuration. Other valid flags might still be important, and an error on one flag should not hinder updates or changes to the flag configuration.
```
Comment (toddbaert) [inline docs/architecture-decisions/flag_configuration_parsing.md:51]: I have to admit, I have a strong preference for failing to start up if the config is invalid in any way, and fully rejecting an update if it's invalid in any way. I think this is what most users expect from invalid configurations. I would be open to a switch that would ignore validation errors, in which case the behavior might not be well-defined (but it would allow us to test some of our edge cases).

I think removing invalid flags actually leads to more confusing behaviors, since flagd will start up but certain flags might just be missing due to a typo. I think it's better to simply not start up if any single flag is invalid. When an update is received, I think similarly - it's better if the entire update fails than a single flag is "silently" removed (only showing up in a log or something).

I also think implementing removal of only invalid flags will be tricky.

I'm really interested in other's opinions here though. Maybe I'm in the minority here.

Comment (aepfli) [reply docs/architecture-decisions/flag_configuration_parsing.md:51]: I would not remove them, I would persist them with a parse_error in the flag storage. this is now implementation specific for eg a hashmap:

- an existing key without a value is a parse_error reason
-  an existing key with a flag, will trigger the resolution process
- a non existing key, is a flag not found

This way, we keep the user informed about possible flag issues, but ensure that updates for other flags are working.

Comment (aepfli) [reply docs/architecture-decisions/flag_configuration_parsing.md:51]: enhanced the proposal with flow diagrams for easier understanding of the proposed flow

**Comment by toddbaert** [comment]
@aepfli sorry, I started typing without selecting the textbox first and accidently closed the issue :sweat_smile: 

To summarize my alternative proposal:

- use https://flagd.dev/schema/v0/flags.json to validate the whole document, on each sync
- at startup, if anything is invalid, flagd fails to start (this is already partially implemented; if we can't parse the flags configuration, so we'd just follow that same path)
- on any sync update, if anything is invalid, the update is ignored and an error/warning message is logged louded (again, this is already partially implemented; if we can't parse the flags configuration, so we'd just follow that same path)

Advantages:

- very easy to implement, just add a validation step in the exact same code-location we already parse
- we can remove many of our "error-edge-cases" in our test suite, since they will never be allowed
- we rely fully on a schema for validation so we don't have to implement any detailed error handling code

Disadvantages:

- 1 bad flag will prevent a whole update - however, this is already true if a single flag cannot be parsed (invalid JSON or nonsense variant configuration, etc) so we can never fully remove this situation, so I think we should just embrace this flow

**Comment by aepfli** [comment]
@toddbaert I also added your proposal.

I think it would be time to start some kind of vote - or give people the chance to add their remark - @open-feature/flagd-triagers please take a look
